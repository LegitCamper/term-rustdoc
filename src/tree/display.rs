use super::*;
use std::fmt::{self, Debug};

/// skip formatting the field when the value is empty or false
macro_rules! skip_fmt {
    ($base:ident, $self:ident . $($field:ident)+ ) => {$(
        if !$self.$field.is_empty() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (bool: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (option: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field.is_some() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (0: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field != 0 {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
}

impl Debug for DModule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DModule");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . modules structs unions enums
            functions traits constants statics type_alias imports macros
        );
        base.finish()
    }
}

impl Debug for DImpl {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DImpl");
        skip_fmt!(base, self . inherent trait_ auto blanket);
        base.finish()
    }
}

impl Debug for DStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DStruct");
        base.field("id", &self.id);
        skip_fmt!(bool: base, self.contain_private_fields);
        skip_fmt!(base, self . fields impls);
        base.finish()
    }
}

impl Debug for DUnion {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DUnion");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . fields impls
        );
        base.finish()
    }
}

impl Debug for DEnum {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DEnum");
        base.field("id", &self.id);
        skip_fmt!(base, self . variants impls);
        base.finish()
    }
}

impl Debug for DTrait {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTrait");
        base.field("id", &self.id);
        skip_fmt!(base, self . types constants functions implementations);
        base.finish()
    }
}

impl Debug for DTypeAlias {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTypeAlias");
        base.field("id", &self.id);
        skip_fmt!(option: base, self.source_path);
        base.finish()
    }
}

impl Debug for TotolCount {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("TotolCount");
        skip_fmt!(
            0: base, self . modules structs unions enums functions
            traits constants statics type_alias imports
            macros_decl macros_func macros_attr macros_derv
        );
        base.finish()
    }
}
