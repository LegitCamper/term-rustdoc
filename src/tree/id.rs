use crate::util::{xformat, CompactStringExt, XString};
use rustdoc_types::{
    Crate, GenericArg, GenericArgs, Id, Item, ItemEnum, ItemKind, ItemSummary, Path, Type,
};
use std::{borrow::Borrow, cell::RefCell, collections::HashMap};

// NOTE: potentially small improvements on id by using CompactString,
// but this requires the HashMaps to be indexed by other id types:
// * IndexMap<CompactString, Value, FxHash> can improve the hashing speed and
//   get the value from multiple type keys via Equivalent trait
// * so does the tiny improvement really matter?
// #[repr(transparent)]
// pub struct ID {
//     pub id: XString,
// }
pub type ID = String;
pub type IDs = Box<[ID]>;

pub trait IdToID: Sized {
    fn to_ID(&self) -> ID;
    fn into_ID(self) -> ID;
}

impl IdToID for Id {
    fn to_ID(&self) -> ID {
        self.0.clone()
    }

    fn into_ID(self) -> ID {
        self.0
    }
}

impl IdToID for String {
    fn to_ID(&self) -> ID {
        self.clone()
    }

    fn into_ID(self) -> ID {
        self
    }
}

impl IdToID for &str {
    fn to_ID(&self) -> ID {
        (*self).to_owned()
    }

    fn into_ID(self) -> ID {
        self.to_owned()
    }
}

pub trait SliceToIds {
    fn to_ids(&self) -> IDs;
}
impl<T: IdToID> SliceToIds for [T] {
    fn to_ids(&self) -> IDs {
        self.iter().map(|id| id.to_ID()).collect()
    }
}

pub trait IdAsStr {
    fn id_str(&self) -> &str;
}
impl IdAsStr for str {
    fn id_str(&self) -> &str {
        self
    }
}
impl IdAsStr for ID {
    fn id_str(&self) -> &str {
        self
    }
}
impl IdAsStr for Id {
    fn id_str(&self) -> &str {
        &self.0
    }
}
impl<T: IdAsStr> IdAsStr for &T {
    fn id_str(&self) -> &str {
        T::id_str(self)
    }
}

// Crate.index
pub type IndexMap = HashMap<Id, Item>;
// Crate.paths
pub type PathMap = HashMap<Id, ItemSummary>;

pub struct IDMap<'krate> {
    pub index: &'krate IndexMap,
    pub paths: &'krate PathMap,
    id_buffer: RefCell<String>,
}

impl IDMap<'_> {
    pub fn from_crate(krate: &Crate) -> IDMap<'_> {
        IDMap {
            index: &krate.index,
            paths: &krate.paths,
            id_buffer: Default::default(),
        }
    }

    /// Use [`rustdoc_types::Id`] in a buffered way in hot querys.
    pub fn use_id<T>(&self, id: &str, f: impl FnOnce(&Id) -> T) -> T {
        // idbuf always serves as Id used in a query
        let mut buf = self.id_buffer.take();
        buf.clear();
        buf.push_str(id);

        let id = Id(buf);
        let val = f(&id);

        // put the buffer back to be used next time
        self.id_buffer.replace(id.0);

        val
    }
}

/// Get the shortest item name only based on IndexMap.
impl IDMap<'_> {
    pub fn get_item(&self, id: &str) -> Option<&Item> {
        self.use_id(id, |id| self.index.get(id))
    }

    // fn use_item(&self, id: &Id, f: impl FnOnce(&Item) -> XString) -> Option<XString> {
    //     self.get_item(&id.0).map(f)
    // }
    //

    /// If the id doesn't refer to an Item, emit a warn and use the id as the result.
    fn use_item_well(&self, id: &str, f: impl FnOnce(&Item) -> XString) -> XString {
        match self.get_item(id).map(f) {
            Some(s) => s,
            None => {
                warn!("Id({id}) doesn't refer to an Item in IndexMap");
                XString::from(id)
            }
        }
    }

    /// * If the id refers to an Item with a name, use the name;
    /// * if not, try getting the name depending on item type;
    /// * id will be used as the result when name parsing is not yet supported.
    pub fn name<S>(&self, id: &S) -> XString
    where
        S: ?Sized + IdAsStr,
    {
        let id = id.id_str();
        self.use_item_well(id, |item| {
            let name = item.name.as_deref().map(XString::from);
            name.unwrap_or_else(|| item_name(item).unwrap_or_else(|| id.into()))
        })
    }
}

/// Deduce the name from its item type.
fn item_name(item: &Item) -> Option<XString> {
    match &item.inner {
        ItemEnum::Impl(item) => {
            let implementor = item
                .blanket_impl
                .as_ref()
                .and_then(type_name)
                .or_else(|| type_name(&item.for_))
                .unwrap_or_default();
            let trait_ = item.trait_.as_ref().and_then(resolved_path_name)?;
            Some(xformat!("{implementor}: {trait_}"))
        }
        _ => None,
    }
}

const COMMA: XString = XString::new_inline(", ");

fn type_name(ty: &Type) -> Option<XString> {
    match ty {
        Type::ResolvedPath(p) => resolved_path_name(p),
        Type::Generic(t) => Some(t.as_str().into()),
        _ => None,
    }
}

fn resolved_path_name(p: &Path) -> Option<XString> {
    let name = p.name.as_str();
    match p.args.as_deref() {
        Some(GenericArgs::AngleBracketed { args, bindings: _ }) => {
            // FIXME: bindings without args
            if args.is_empty() {
                Some(name.into())
            } else {
                Some(xformat!(
                    "{name}<{}>",
                    args.iter()
                        .filter_map(generic_arg_name)
                        .intersperse(COMMA)
                        .collect::<XString>()
                ))
            }
        }
        Some(GenericArgs::Parenthesized { inputs, output }) => {
            let args = inputs
                .iter()
                .filter_map(type_name)
                .intersperse(COMMA)
                .collect::<XString>();
            let ret = output
                .as_ref()
                .and_then(|t| Some(xformat!(" -> {}", type_name(t)?)))
                .unwrap_or_default();
            Some(xformat!("{name}({args}){ret}"))
        }
        None => Some(name.into()),
    }
}

fn generic_arg_name(arg: &GenericArg) -> Option<XString> {
    match arg {
        GenericArg::Lifetime(life) => Some(life.as_str().into()),
        GenericArg::Type(ty) => type_name(ty),
        GenericArg::Const(_) => None,
        GenericArg::Infer => Some(XString::new_inline("_")),
    }
}

/// Get the external item path only based on PathMap.
impl IDMap<'_> {
    pub fn get_path(&self, id: &str) -> Option<&ItemSummary> {
        self.use_id(id, |id| self.paths.get(id))
    }

    // fn use_path(&self, id: &Id, f: impl FnOnce(&ItemSummary) -> XString) -> Option<XString> {
    //     self.get_path(&id.0).map(f)
    // }

    /// If the id doesn't refer to an ItemSummary, emit a warn and use the id as the result.
    fn use_path_well(&self, id: &str, f: impl FnOnce(&ItemSummary) -> XString) -> XString {
        match self.get_path(id).map(f) {
            Some(s) => s,
            None => {
                warn!("Id({id}) doesn't refer to an Item in IndexMap");
                XString::from(id)
            }
        }
    }

    /// If the id doesn't refer to an ItemSummary with exact given kind,
    /// emit a warn and use the id as the result.
    pub fn path<S, K>(&self, id: &S, kind: K) -> XString
    where
        S: ?Sized + IdAsStr,
        K: Borrow<ItemKind>,
    {
        self.use_path_well(id.id_str(), move |item| {
            let id = id.id_str();
            let kind = kind.borrow();
            if &item.kind == kind {
                item.path.join_compact("::")
            } else {
                warn!(
                    "Id({id}) in PathMap is found as {:?}, but {kind:?} is required",
                    item.kind
                );
                XString::from(id)
            }
        })
    }
}
