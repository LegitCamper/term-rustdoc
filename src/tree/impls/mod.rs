mod debug;

#[macro_use]
pub mod show;
