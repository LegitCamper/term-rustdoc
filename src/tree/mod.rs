use rustdoc_types::*;
use std::collections::HashMap;

mod display;

#[allow(non_snake_case)]
mod id;
pub use id::{IDs, IdToID, SliceToIds, ID};

pub type Arr<T> = Box<[T]>;
/// Use `IDs::default()` instead of `IDs::new()` to create an empty IDs.
// pub type Vec<T> = std::sync::Arc<[T]>;
// pub type Arc<T> = std::sync::Arc<T>;

// Crate.index
pub type IndexMap = HashMap<Id, Item>;
// Crate.paths
pub type PathMap = HashMap<Id, ItemSummary>;

/// Module tree with structural items.
/// All the items only carry ids without actual data.
// NOTE: small improvement by turning all the types of fields
// from Vec to Arr after instantiation.
#[derive(Default)]
pub struct DModule {
    pub id: ID,
    // If true, this module is not part of the public API,
    // but it contains items that are re-exported as public API.
    // is_stripped: bool,
    pub modules: Vec<DModule>,
    pub structs: Vec<DStruct>,
    pub unions: Vec<DUnion>,
    pub enums: Vec<DEnum>,
    pub functions: Vec<DFunctions>,
    pub traits: Vec<DTrait>,
    pub constants: Vec<DConstant>,
    pub statics: Vec<DStatic>,
    pub type_alias: Vec<DTypeAlias>,
    pub imports: Vec<Import>,
    pub macros: Vec<DMacro>,
}

impl DModule {
    pub fn new(doc: &Crate) -> Self {
        // root module/crate name
        let (id, root) = doc
            .index
            .iter()
            .find_map(|(id, item)| {
                if item.crate_id == 0 {
                    if let ItemEnum::Module(Module {
                        is_crate: true,
                        items,
                        ..
                    }) = &item.inner
                    {
                        return Some((id.clone(), items.as_slice()));
                    }
                }
                None
            })
            .expect("root module not found");
        Self::new_inner(id.to_ID(), root, &doc.index)
    }

    fn new_inner(id: ID, inner_items: &[Id], index: &IndexMap) -> Self {
        let mut dmod = DModule {
            id,
            ..Default::default()
        };
        dmod.extract_items(inner_items, index);
        dmod
    }

    fn extract_items(&mut self, inner_items: &[Id], index: &IndexMap) {
        for item_id in inner_items {
            match index.get(item_id) {
                Some(item) => self.append(item, index),
                None => warn!("the local item {item_id:?} not found in Crate's index"),
            }
        }
    }

    fn append(&mut self, item: &Item, index: &IndexMap) {
        use ItemEnum::*;
        let id = item.id.to_ID();
        match &item.inner {
            Module(item) => self.modules.push(Self::new_inner(id, &item.items, index)),
            Struct(item) => self.structs.push(DStruct::new(id, item, index)),
            Union(item) => self.unions.push(DUnion::new(id, item, index)),
            Enum(item) => self.enums.push(DEnum::new(id, item, index)),
            Function(_) => self.functions.push(DFunctions::new(id)),
            Trait(item) => self.traits.push(DTrait::new(id, item, index)),
            Constant(_) => self.constants.push(DConstant::new(id)),
            Static(_) => self.statics.push(DStatic::new(id)),
            TypeAlias(item) => self.type_alias.push(DTypeAlias::new(id, item)),
            Macro(_) => self.macros.push(DMacro::new(id, DMacroKind::Declarative)),
            ProcMacro(proc) => self.macros.push(DMacro::new(
                id,
                match proc.kind {
                    MacroKind::Bang => DMacroKind::ProcFunction,
                    MacroKind::Attr => DMacroKind::ProcAttribute,
                    MacroKind::Derive => DMacroKind::ProcDerive,
                },
            )),
            // Primitive(_) => todo!(),
            _ => (),
        }
    }
}

#[derive(Default)]
pub struct DImpl {
    pub inherent: IDs,
    pub trait_: IDs,
    pub auto: IDs,
    pub blanket: IDs,
}
impl DImpl {
    fn new(ids: &[Id], index: &IndexMap) -> Box<Self> {
        if ids.is_empty() {
            return Default::default();
        }
        let [mut inherent, mut trait_, mut auto, mut blanket]: [Vec<ID>; 4] = Default::default();
        for Id(id) in ids {
            if id.starts_with("a:") {
                auto.push(id.to_ID());
            } else if id.starts_with("b:") {
                blanket.push(id.to_ID());
            } else {
                let id = Id(id.clone());
                if let Some(item) = index.get(&id) {
                    if let ItemEnum::Impl(impl_) = &item.inner {
                        if impl_.trait_.is_none() {
                            inherent.push(id.into_ID());
                        } else {
                            trait_.push(id.into_ID());
                        }
                    } else {
                        warn!("{id:?} in Crate's index doesn't refer to an impl item");
                    }
                } else {
                    warn!("the impl with {id:?} not found in Crate's index");
                }
            }
        }
        Box::new(DImpl {
            inherent: inherent.into(),
            trait_: trait_.into(),
            auto: auto.into(),
            blanket: blanket.into(),
        })
    }
    fn is_empty(&self) -> bool {
        self.inherent.is_empty()
            && self.trait_.is_empty()
            && self.auto.is_empty()
            && self.blanket.is_empty()
    }
}

// TODO:
// *  structs, enums, and enum variants: [non_exhaustive]
//
// [non_exhaustive]: https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute

pub struct DStruct {
    pub id: ID,
    pub fields: IDs,
    pub contain_private_fields: bool,
    pub impls: Box<DImpl>,
}
impl DStruct {
    fn new(id: ID, item: &Struct, index: &IndexMap) -> Self {
        let mut contain_private_fields = false;
        let fields = match &item.kind {
            StructKind::Unit => IDs::default(),
            StructKind::Tuple(fields) => fields
                .iter()
                .filter_map(|f| {
                    let id = f.as_ref().map(|id| id.to_ID());
                    if id.is_none() {
                        contain_private_fields = true;
                    }
                    id
                })
                .collect(),
            StructKind::Plain {
                fields,
                fields_stripped,
            } => {
                contain_private_fields = *fields_stripped;
                fields.to_ids()
            }
        };
        let impls = DImpl::new(&item.impls, index);
        DStruct {
            id,
            fields,
            contain_private_fields,
            impls,
        }
    }
}

pub struct DUnion {
    pub id: ID,
    pub fields: IDs,
    pub impls: Box<DImpl>,
}
impl DUnion {
    fn new(id: ID, item: &Union, index: &IndexMap) -> Self {
        DUnion {
            id,
            fields: item.fields.to_ids(),
            impls: DImpl::new(&item.impls, index),
        }
    }
}

pub struct DEnum {
    pub id: ID,
    pub variants: IDs,
    // variants_stripped: bool, -> Does this really make sense?
    pub impls: Box<DImpl>,
}
impl DEnum {
    fn new(id: ID, item: &Enum, index: &IndexMap) -> Self {
        DEnum {
            id,
            variants: item.variants.to_ids(),
            impls: DImpl::new(&item.impls, index),
        }
    }
}

pub struct DTrait {
    pub id: ID,
    pub types: IDs,
    pub constants: IDs,
    pub functions: IDs,
    pub implementations: IDs,
}
impl DTrait {
    fn new(id: ID, item: &Trait, index: &IndexMap) -> Self {
        let [mut types, mut constants, mut functions]: [Vec<ID>; 3] = Default::default();
        let trait_id = &id;
        for id in &item.items {
            if let Some(assoc) = index.get(id) {
                let id = id.to_ID(); // id == assoc.id
                match &assoc.inner {
                    ItemEnum::AssocType { .. } => types.push(id),
                    ItemEnum::AssocConst { .. } => constants.push(id),
                    ItemEnum::Function(_) => functions.push(id),
                    _ => warn!(
                        "`{id}` should refer to an associated item \
                         (type/constant/function) in Trait `{trait_id}`"
                    ),
                }
            } else {
                warn!("the trait item {id:?} not found in Crate's index");
            }
        }
        DTrait {
            id,
            types: types.into(),
            constants: constants.into(),
            functions: functions.into(),
            implementations: item.implementations.to_ids(),
        }
    }
}

/// generate id wrapper types for simple items
macro_rules! gen_simple_items {
    ($($name:ident),+) => {$(
        #[derive(Debug)] pub struct $name { pub id: ID, }
        impl $name { fn new(id: ID) -> Self { Self { id } } }
    )+};
}
gen_simple_items!(DFunctions, DStatic, DConstant);

pub struct DTypeAlias {
    pub id: ID,
    /// points to a source type resolved as path
    ///
    /// A type alias may point to non-path-based type though.
    pub source_path: Option<ID>,
}
impl DTypeAlias {
    pub fn new(id: ID, item: &TypeAlias) -> Self {
        Self {
            id,
            source_path: match &item.type_ {
                Type::ResolvedPath(Path { id, .. }) => Some(id.to_ID()),
                _ => None,
            },
        }
    }
}

#[derive(Debug)]
pub enum DMacroKind {
    Declarative,
    ProcFunction,
    ProcAttribute,
    ProcDerive,
}
#[derive(Debug)]
pub struct DMacro {
    pub id: ID,
    pub kind: DMacroKind,
}
impl DMacro {
    pub fn new(id: ID, kind: DMacroKind) -> Self {
        Self { id, kind }
    }
}
