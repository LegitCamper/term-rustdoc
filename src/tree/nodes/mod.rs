mod impls;
use std::ops::Not;

pub use impls::DImpl;

mod structs;
pub use structs::DStruct;

mod unions;
pub use unions::DUnion;

mod enums;
pub use enums::DEnum;

mod traits;
pub use traits::DTrait;

mod type_alias;
pub use type_alias::DTypeAlias;

mod imports;

use super::IDMap;
use crate::tree::{
    impls::show::{DocTree, Show},
    IdToID, IndexMap, ID,
};
use rustdoc_types::{Crate, Id, Item, ItemEnum, ItemKind, MacroKind, Module};

/// Module tree with structural items.
/// All the items only carry ids without actual data.
// NOTE: small improvement by turning all the types of fields
// from Vec to Arr after instantiation.
#[derive(Default)]
pub struct DModule {
    pub id: ID,
    // If true, this module is not part of the public API,
    // but it contains items that are re-exported as public API.
    // is_stripped: bool,
    pub modules: Vec<DModule>,
    pub structs: Vec<DStruct>,
    pub unions: Vec<DUnion>,
    pub enums: Vec<DEnum>,
    pub functions: Vec<DFunction>,
    pub traits: Vec<DTrait>,
    pub constants: Vec<DConstant>,
    pub statics: Vec<DStatic>,
    pub type_alias: Vec<DTypeAlias>,
    pub imports: Vec<()>,
    pub macros_decl: Vec<DMacroDecl>,
    pub macros_func: Vec<DMacroFunc>,
    pub macros_attr: Vec<DMacroAttr>,
    pub macros_derv: Vec<DMacroDerv>,
}

impl DModule {
    pub fn new(doc: &Crate) -> Self {
        // root module/crate name
        let (id, root) = doc
            .index
            .iter()
            .find_map(|(id, item)| {
                if item.crate_id == 0 {
                    if let ItemEnum::Module(Module {
                        is_crate: true,
                        items,
                        ..
                    }) = &item.inner
                    {
                        return Some((id.to_ID(), items.as_slice()));
                    }
                }
                None
            })
            .expect("root module not found");
        Self::new_inner(id, root, &doc.index)
    }

    fn new_inner(id: ID, inner_items: &[Id], index: &IndexMap) -> Self {
        let mut dmod = DModule {
            id,
            ..Default::default()
        };
        dmod.extract_items(inner_items, index);
        dmod
    }

    fn extract_items(&mut self, inner_items: &[Id], index: &IndexMap) {
        for item_id in inner_items {
            match index.get(item_id) {
                Some(item) => self.append(item, index),
                None => warn!("the local item {item_id:?} not found in Crate's index"),
            }
        }
    }

    fn append(&mut self, item: &Item, index: &IndexMap) {
        use ItemEnum::*;
        let id = item.id.to_ID();
        match &item.inner {
            Module(item) => self.modules.push(Self::new_inner(id, &item.items, index)),
            Struct(item) => self.structs.push(DStruct::new(id, item, index)),
            Union(item) => self.unions.push(DUnion::new(id, item, index)),
            Enum(item) => self.enums.push(DEnum::new(id, item, index)),
            Function(_) => self.functions.push(DFunction::new(id)),
            Trait(item) => self.traits.push(DTrait::new(id, item, index)),
            Constant(_) => self.constants.push(DConstant::new(id)),
            Static(_) => self.statics.push(DStatic::new(id)),
            TypeAlias(item) => self.type_alias.push(DTypeAlias::new(id, item)),
            Macro(_) => self.macros_decl.push(DMacroDecl::new(id)),
            ProcMacro(proc) => match proc.kind {
                MacroKind::Bang => self.macros_func.push(DMacroFunc::new(id)),
                MacroKind::Attr => self.macros_attr.push(DMacroAttr::new(id)),
                MacroKind::Derive => self.macros_derv.push(DMacroDerv::new(id)),
            },
            // Primitive(_) => todo!(),
            _ => (),
        }
    }
}

macro_rules! chain {
    ($field:ident, $node:literal, $self:ident, $map:ident) => {
        $self.$field.is_empty().not().then(|| {
            $node
                .show()
                .with_leaves($self.$field.iter().map(|val| val.show_prettier($map)))
        })
    };
}

/// To a recursive tree displayed with ids as nodes.
impl Show for DModule {
    fn show(&self) -> DocTree {
        format!("[mod] {}", self.id).show().with_leaves(
            self.modules.iter().map(DModule::show).chain([
                "Structs"
                    .show()
                    .with_leaves(self.structs.iter().map(DStruct::show)),
                "Unions"
                    .show()
                    .with_leaves(self.unions.iter().map(DUnion::show)),
                "Enums"
                    .show()
                    .with_leaves(self.enums.iter().map(DEnum::show)),
                "Traits"
                    .show()
                    .with_leaves(self.traits.iter().map(DTrait::show)),
                "Functions"
                    .show()
                    .with_leaves(self.functions.iter().map(DFunction::show)),
                "Constants"
                    .show()
                    .with_leaves(self.constants.iter().map(DConstant::show)),
                "Statics"
                    .show()
                    .with_leaves(self.statics.iter().map(DStatic::show)),
                "Macros - Declarative"
                    .show()
                    .with_leaves(self.macros_decl.iter().map(DMacroDecl::show)),
                "Macro - Function"
                    .show()
                    .with_leaves(self.macros_func.iter().map(DMacroFunc::show)),
                "Macro - Attribute"
                    .show()
                    .with_leaves(self.macros_attr.iter().map(DMacroAttr::show)),
                "Macro - Derive"
                    .show()
                    .with_leaves(self.macros_derv.iter().map(DMacroDerv::show)),
            ]),
        )
    }

    fn show_prettier(&self, map: &IDMap) -> DocTree {
        node!("[mod] {}", map.path(&self.id, ItemKind::Module)).with_leaves(
            self.modules
                .iter()
                .map(|m| m.show_prettier(map))
                .chain(chain!(structs, "Structs", self, map))
                .chain(chain!(unions, "Unions", self, map))
                .chain(chain!(enums, "Enums", self, map))
                .chain(chain!(traits, "Traits", self, map))
                .chain(chain!(functions, "Functions", self, map))
                .chain(chain!(constants, "Constants", self, map))
                .chain(chain!(statics, "Statics", self, map))
                .chain(chain!(macros_decl, "Macros - Declarative", self, map))
                .chain(chain!(macros_func, "Macros - Function", self, map))
                .chain(chain!(macros_attr, "Macros - Attribute", self, map))
                .chain(chain!(macros_derv, "Macros - Derive", self, map)),
        )
    }
}

/// generate id wrapper types for simple items
macro_rules! gen_simple_items {
    ($($name:ident => $show:literal -> $kind:ident),+ $(,)?) => {$(
        #[derive(Debug)] pub struct $name { pub id: ID, }
        impl $name { pub fn new(id: ID) -> Self { Self { id } } }
        impl Show for $name {
            fn show(&self) -> DocTree { self.id.show() }
            fn show_prettier(&self, map: &IDMap) -> DocTree {
                node!($show, map.path(&self.id, ItemKind::$kind))
            }
        }
    )+};
}
gen_simple_items! {
    DFunction => "[fn] {}" -> Function ,
    DStatic => "[static] {}" -> Static,
    DConstant => "[constant] {}" -> Constant,
    DMacroDecl => "[macro decl] {}" -> Macro,
    DMacroFunc => "[macro func] {}" -> Macro,
    DMacroAttr => "[macro attr] {}" -> ProcAttribute,
    DMacroDerv => "[macro derv] {}" -> ProcDerive,
}

// TODO:
// *  structs, enums, and enum variants: [non_exhaustive]
//
// [non_exhaustive]: https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute
