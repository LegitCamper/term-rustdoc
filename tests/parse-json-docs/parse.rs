use crate::{doc, shot, snap};
use similar_asserts::assert_eq;
use term_rustdoc::tree::{Show, TreeLines};

#[test]
fn parse_module() {
    let (treelines, empty) = TreeLines::new_with(doc(), |doc| doc.dmodule_show_prettier());
    let doc = treelines.doc();
    let dmod = doc.dmodule();
    snap!("DModule", dmod);
    shot!("show-id", dmod.show());

    let tree = doc.dmodule_show_prettier();
    let display = tree.to_string();
    shot!("show-prettier", display);

    let display_new = treelines.display_as_plain_text();
    assert_eq!(expected: display, actual: display_new);
    snap!("flatten-tree", treelines.all_lines());
    shot!("empty-tree-with-same-depth", empty);

    // item tree
    shot!("item-tree", dmod.item_tree(&doc));

    snap!(dmod.current_items_counts(), @r###"
    ItemCount {
        modules: 1,
        structs: 2,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);
    snap!(dmod.recursive_items_counts(), @r###"
    ItemCount {
        modules: 2,
        structs: 2,
        enums: 1,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);

    snap!(dmod.current_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 3,
            structs: 3,
        },
        inherent: ImplCount {
            kind: Inherent,
            total: 1,
            structs: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 2,
            structs: 2,
        },
    }
    "###);
    snap!(dmod.recursive_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 5,
            structs: 3,
            enums: 2,
        },
        inherent: ImplCount {
            kind: Both,
            total: 2,
            structs: 1,
            enums: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 3,
            structs: 2,
            enums: 1,
        },
    }
    "###);

    // struct inner
    let (struct_, _) = TreeLines::new_with(treelines.doc(), |doc| {
        doc.dmodule().structs[0].show_prettier(doc)
    });
    shot!(struct_.display_as_plain_text(), @r###"
    integration::AUnitStruct
    ├── No Fields!
    └── Implementations
        ├── Trait Impls
        │   └── AUnitStruct: ATrait
        ├── Auto Impls
        │   ├── AUnitStruct: RefUnwindSafe
        │   ├── AUnitStruct: Send
        │   ├── AUnitStruct: Sync
        │   ├── AUnitStruct: Unpin
        │   └── AUnitStruct: UnwindSafe
        └── Blanket Impls
            ├── T: Any
            ├── T: Borrow<T>
            ├── T: BorrowMut<T>
            ├── T: From<T>
            ├── T: Into<U>
            ├── T: TryFrom<U>
            └── T: TryInto<U>
    "###);
}
