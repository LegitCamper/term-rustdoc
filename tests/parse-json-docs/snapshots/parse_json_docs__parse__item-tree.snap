---
source: tests/parse-json-docs/parse.rs
expression: dmod.item_tree(&doc)
---
integration
├── [Fn] func_dyn_trait
├── [Fn] func_with_1arg
├── [Fn] func_with_1arg_and_ret
├── [Fn] func_with_no_args
├── [Const] ACONSTANT
├── [Const] ASTATIC
├── [macro decl] a_decl_macro
├── [Trait] ATrait
├── [Struct] AUnitStruct
├── [Struct] FieldsNamedStruct
└── [Mod] integration::submod1
    ├── [Enum] AUnitEnum
    └── [Mod] integration::submod1::submod2
        ├── [Trait] ATraitNeverImplementedForTypes
        ├── [Struct] AStructAlias
        └── [Struct] AUnitStruct
